core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java

```diff
@@ -0,0 +1,279 @@
+import io.trino.sql.tree.SimpleCaseExpression;
+import io.trino.sql.tree.SymbolReference;
+import io.trino.sql.tree.WhenClause;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static io.trino.SystemSessionProperties.isOptimizeCaseExpressionPredicate;
+import static io.trino.sql.planner.plan.Patterns.filter;
+import static io.trino.sql.tree.ComparisonExpression.Operator.EQUAL;
+
+/**
+ * Re-writes CASE expression predicate into a series of OR clauses
+ * <p>
+ * (CASE
+ * WHEN expression=constant1 THEN result1
+ * WHEN expression=constant2 THEN result2
+ * ELSE elseResult
+ * END) = value
+ * <p>
+ * becomes
+ * <p>
+ * (result1 = value AND expression=constant1) OR
+ * (result2 = value AND expression=constant2) OR
+ * (elseResult = value AND !(expression=constant1) AND !(expression=constant2))
+ * <p>
+ * This conversion happens when all of the following criteria are met:
+ * 1. Value is either a constant or a symbol
+ * 2. The LHS expression in all WHEN clauses are the same.
+ *    For example, if one WHEN clause has a expression using col1 and another using col2, it will not work
+ * 3. The relational operator in the WHEN clause is equals. With other operators it is harder to check for exclusivity.
+ * 4. All the RHS expressions are a constant and unique
+ * <p>
+ * This conversion is done so that it is easy for the SimplifyExpressions & other Optimizers to further
+ * simplify this and construct a domain for the column that can be used by Readers .
+ * i.e, SimplifyExpressions can discard all conditions in which result != value and
+ * DomainTranslator can construct a Domain for the column
+ */
+public class RewriteCaseExpressionPredicate
+        implements Rule<FilterNode>
+{
+    private static final Pattern<FilterNode> PATTERN = filter();
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isOptimizeCaseExpressionPredicate(session);
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Rule.Result apply(FilterNode filterNode, Captures captures, Rule.Context context)
+    {
+        Expression rewritten = ExpressionTreeRewriter.rewriteWith(new Rewriter(), filterNode.getPredicate());
+        if (filterNode.getPredicate().equals(rewritten)) {
+            return Result.empty();
+        }
+        return Result.ofPlanNode(new FilterNode(filterNode.getId(), filterNode.getSource(), rewritten));
+    }
+
+    private static class Rewriter
+            extends ExpressionRewriter<Void>
+    {
+        @Override
+        public Expression rewriteComparisonExpression(ComparisonExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (isCaseExpression(node.getLeft()) && isSimpleExpression(node.getRight())) {
```

Comment:
Rewrite the sub-expressions recursively.

Reply:
Done.